### ISSUE TITLE
Exception resolving root view-model when using Autofac child lifetime scopes

### ISSUE THREAD

**Issue by nathan-alden-hp** [canton7/Stylet#109]
I am building a WPF application that loads views and view models from plugin assemblies (called *modules*) at runtime. Due to the structure of my code, I am unable to use a single Autofac lifetime scope. I create a child lifetime scope where I register components dependent on views, view-models, or Stylet:
```cs
public class AutofacBootstrapper<TRootViewModel> : BootstrapperBase
    where TRootViewModel : class
{
    private ILifetimeScope? _lifetimeScope;
    private TRootViewModel? _rootViewModel;

    public sealed override object GetInstance(Type type)
    {
        return _lifetimeScope!.Resolve(type);
    }

    public override void Dispose()
    {
        ScreenExtensions.TryDispose(_rootViewModel);

        _lifetimeScope?.Dispose();

        base.Dispose();
    }

    protected sealed override void ConfigureBootstrapper()
    {
        var containerBuilder = new ContainerBuilder();

        // Register components not dependent on views, view-models, or Stylet
        OnRegisterComponents(containerBuilder);

        IContainer container = containerBuilder.Build();

        _lifetimeScope = container.BeginLifetimeScope(
            scopedContainerBuilder =>
            {
                // Register host views and view-models

                scopedContainerBuilder.RegisterType<NavigationItemView>().AsSelf().ExternallyOwned();
                scopedContainerBuilder.RegisterType<NavigationItemViewModel>().AsSelf().ExternallyOwned();
                scopedContainerBuilder.RegisterType<RootView>().AsSelf().ExternallyOwned();
                scopedContainerBuilder.RegisterType<RootViewModel>().AsSelf().ExternallyOwned();

                scopedContainerBuilder.RegisterType<ModuleFactory>().As<IModuleFactory>().SingleInstance();

                // Load module assemblies

                var moduleAssemblyLoader = container.Resolve<IModuleAssemblyLoader>();
                IReadOnlyCollection<string> moduleAssemblyPaths = moduleAssemblyLoader.FindModuleAssemblyPaths($"*.{IModuleAssembly.FilenameSuffix}.dll");

                moduleAssemblyLoader.LoadModuleAssemblies(container, moduleAssemblyPaths);

                // Register module assembly components

                foreach (IModuleAssembly moduleAssembly in moduleAssemblyLoader.LoadedModuleAssemblies)
                {
                    moduleAssembly.RegisterComponents(scopedContainerBuilder);
                }

                // Register modules

                foreach (Assembly assembly in moduleAssemblyLoader.LoadedAssemblies)
                {
                    scopedContainerBuilder
                        .RegisterAssemblyTypes(assembly)
                        .Where(type => type.IsAssignableTo<IModule>() && type.IsPublic && !type.IsAbstract)
                        .AsSelf();
                }

                // Register Stylet components

                var viewManagerConfig =
                    new ViewManagerConfig
                    {
                        ViewFactory = GetInstance,
                        ViewAssemblies = new List<Assembly> { typeof(App).Assembly }
                    };

                viewManagerConfig.ViewAssemblies.AddRange(moduleAssemblyLoader.LoadedAssemblies);

                scopedContainerBuilder.RegisterType<MessageBoxViewModel>().As<IMessageBoxViewModel>().ExternallyOwned();
                scopedContainerBuilder.RegisterInstance<IViewManager>(new SkinAwareViewManager(container.Resolve<IHostConfiguration>(), viewManagerConfig));
                scopedContainerBuilder.RegisterType<WindowManager>().As<IWindowManager>().SingleInstance();
                scopedContainerBuilder.RegisterInstance<IWindowManagerConfig>(this).ExternallyOwned();
            });
    }

    protected override void Launch()
    {
        DisplayRootView(_rootViewModel = (TRootViewModel)GetInstance(typeof(TRootViewModel)));
    }

    protected override void OnExit(ExitEventArgs e)
    {
        _lifetimeScope?.Dispose();
    }

    protected virtual void OnRegisterComponents(ContainerBuilder containerBuilder)
    {
    }

    protected T GetInstance<T>()
    {
        return (T)GetInstance(typeof(T));
    }
}
```
You can see how I load module assemblies and then pass them to the `ViewManagerConfig` at the appropriate time. Unfortunately, I get an exception inside `DisplayRootView` (note that `TRootViewModel` is a view-model in the main application assembly, *not* a module assembly):
```
System.InvalidOperationException: 'The ViewManager resource is unassigned. This should have been set by the Bootstrapper'
```
![image](https://user-images.githubusercontent.com/40613229/75905997-eeae2080-5e13-11ea-8787-7cc53af7c13b.png)
![image](https://user-images.githubusercontent.com/40613229/75907893-24083d80-5e17-11ea-9cfa-bd43a8058bb3.png)
Something important to note: This exception only occurs when at least one view-model in a module assembly derives from `Screen` (which I require). If I remove the derivation, the exception no longer occurs.

I took a look at the Stylet source code and found the place where the exception is thrown, but there's so much magic going on that I haven't been able to determine the cause or how to fix it. Do you have any advice for my scenario? If there were a way to change the view assemblies list after DI registration then I wouldn't have to use the child lifetime scope.

**Comment by canton7** [comment]
So on the face of it, it looks like [this line](https://github.com/canton7/Stylet/blob/1ef62f568d41502cbca9770dec8326f9cf1c37c7/Stylet/BootstrapperBase.cs#L78) isn't being executed, or `this.GetInstance(typeof(IViewManager))` is returning `null`. You can investigate by putting a breakpoint in your `GetInstance` override.

It might also be that [this resolution](https://github.com/canton7/Stylet/blob/1ef62f568d41502cbca9770dec8326f9cf1c37c7/Stylet/Xaml/View.cs#L79) is failing, because the view isn't property associated with the Application in some way. I'm not sure what could cause that, though.

Start by looking at whether `BootstrapperBase` is trying to resolve `IViewManager`, and whether it's getting the right thing back.

If you weren't aware, there's a working Autofac bootstrapper [here](https://github.com/canton7/Stylet/blob/develop/Bootstrappers/AutofacBootstrapper.cs), which might work as a starting point.

**Comment by nathan-alden-hp** [comment]
FWIW, I based my Autofac bootstrapper off the one in the repo. I'll take a look at your suggestions. Great library, BTW--much cleaner than Caliburn Micro. 👍

**Comment by nathan-alden-hp** [comment]
Root view model is resolving successfully:
![image](https://user-images.githubusercontent.com/40613229/75918811-68e99f80-5e2a-11ea-99a3-76f135e80e67.png)
The line you linked to does seem to be being executed:
![image](https://user-images.githubusercontent.com/40613229/75919140-00e78900-5e2b-11ea-8a3c-4faca5e99ee4.png)
I think the only choice I have now is to grab the source and import it into my project temporarily for debugging symbols.

**Comment by canton7** [comment]
That suggests that `Start` isn't being called in `BootstrapperBase`, then.

The NuGet package uses SourceLink, which should just work out of the box on recentish versions of VS (you're using NRT's, so VS2019, which is fine). [See here](https://github.com/canton7/Stylet#symbols).

**Comment by nathan-alden-hp** [comment]
Awesome. I just enabled NuGet.org Source Servers and that didn't work, of course. Good on you for enabling complex debugging scenarios!

**Comment by canton7** [comment]
Note that this issue is separate to your root VM resolving. Stylet should be trying to resolve the `IViewManager` during startup, and add it as an application resource. At some point later it's trying to handle a `s:View.Model="{Binding Foo}"`, and that's the point at which it's trying to retrieve the `IViewManager` from the application, but it's not finding it.

**Comment by nathan-alden-hp** [comment]
I suspect I know what the issue likely is. Since the problem only started occurring once I derived `Screen` in my module view-models, it's possible that the way I'm resolving the view-model--currently, directly from the lifetime scope--is incorrect. I'm guessing I'm supposed to go to `IViewManager` to get instances of view-models and views.

**Comment by canton7** [comment]
I can't see how that would make a difference, honestly. `IViewManager` doesn't create VM instances: it's just responsible mainly for binding views to VMs. The default `ViewManager` delegates straight back to the Bootstrapper when it wants to create a view instance.

But the problem is that the `IViewManager` isn't being set as a resouce on the Application. From what you said earlier, the BootstapperBase doesn't even appear to be attempting to do this. So debugging the BootstrapperBase I think is the next step. Debug [this line](https://github.com/canton7/Stylet/blob/1ef62f568d41502cbca9770dec8326f9cf1c37c7/Stylet/BootstrapperBase.cs#L78), see what's wrong.

**Comment by nathan-alden-hp** [comment]
I debugged that line and the `IViewManager` reference is being successfully created, returned, and added to the resource dictionary (as I showed in one of my previous screenshots). Is it possible that my `App.xaml` is causing this issue?
```xaml
<Application x:Class="WpfMultiModule.Host.App"
             xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
             xmlns:s="https://github.com/canton7/Stylet"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:host="clr-namespace:WpfMultiModule.Host">
  <Application.Resources>
    <ResourceDictionary>
      <ResourceDictionary.MergedDictionaries>
        <ResourceDictionary Source="pack://application:,,,/WpfMultiModule.Skins;component/Base/Fonts.xaml" />
        <host:SkinResourceDictionary Skin1Source="pack://application:,,,/WpfMultiModule.Skins;component/Skin1.xaml"
                                           Skin2Source="pack://application:,,,/WpfMultiModule.Skins;component/Skin2.xaml"
                                           Source="pack://application:,,,/WpfMultiModule.Skins;component/Skin1.xaml" />
        <s:ApplicationLoader>
          <s:ApplicationLoader.Bootstrapper>
            <host:Bootstrapper />
          </s:ApplicationLoader.Bootstrapper>
        </s:ApplicationLoader>
      </ResourceDictionary.MergedDictionaries>
    </ResourceDictionary>
  </Application.Resources>
</Application>
```
Is there a specific ordering I need to consider? Does the bootstrapper not work properly in a merged dictionary? I'm going to try reducing my `App.xaml` down to the minimum for testing.

**Comment by canton7** [comment]
The screenshots you shows didn't show the line I was interested in?

Your App.xaml looks fine.

Can you check that the `Application` property on BootstrapperBase is the same instance as `Application.Current` when the exception is thrown from `View`?

**Comment by canton7** [comment]
Also, where exactly are you doing `s:View.Model="..."`? Perhaps the element that's called on isn't part of the view hierarchy for some reason?

**Comment by nathan-alden-hp** [comment]
Funny, as you asked that, I was preparing my next response. I've narrowed down to the issue to my use of `s:View.Model` in a `ContentControl` in `RootView`:
```xaml
<Window x:Class="WpfMultiModule.Host.Views.Root.RootView"
        xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:s="https://github.com/canton7/Stylet"
        xmlns:root="clr-namespace:WpfMultiModule.Host.Views.Root"
        mc:Ignorable="d"
        d:DataContext="{d:DesignInstance root:RootViewModel, IsDesignTimeCreatable=True}"
        Width="800" Height="450">
  <Grid Background="{StaticResource TitleBackgroundBrush}">
    <Grid.RowDefinitions>
      <RowDefinition Height="Auto" />
      <RowDefinition Height="*" />
    </Grid.RowDefinitions>
    <StackPanel Grid.Row="0" Orientation="Horizontal" UseLayoutRounding="True">
      <Image Source="{StaticResource MainLogoImage}" Width="50" Height="50" Margin="10,10,5,10" />
      <TextBlock VerticalAlignment="Center" Text="{Binding Title}" Foreground="{StaticResource TitleForegroundBrush}" FontSize="36" />
    </StackPanel>
    <Grid Grid.Row="1" Background="{StaticResource WindowBackgroundBrush}" UseLayoutRounding="True">
      <Grid.ColumnDefinitions>
        <ColumnDefinition Width="Auto" MinWidth="200" />
        <ColumnDefinition />
      </Grid.ColumnDefinitions>
      <ItemsControl Grid.Column="0" ItemsSource="{Binding Modules}">
        <ItemsControl.ItemsPanel>
          <ItemsPanelTemplate>
            <StackPanel Orientation="Vertical" Background="{StaticResource NavigationBackgroundBrush}" />
          </ItemsPanelTemplate>
        </ItemsControl.ItemsPanel>
        <ItemsControl.ItemTemplate>
          <DataTemplate>
            <StackPanel Orientation="Vertical">
              <ContentControl s:View.Model="{Binding RelativeSource={RelativeSource Mode=FindAncestor, AncestorType=StackPanel}, Path=DataContext}" Cursor="Hand" />
              <Separator Background="{StaticResource NavigationSeparatorBackgroundBrush}" Visibility="{Binding Visibility}" Padding="0" Margin="0" />
            </StackPanel>
          </DataTemplate>
        </ItemsControl.ItemTemplate>
      </ItemsControl>
      <ContentControl Grid.Column="1" Content="{Binding ModuleContent}" />
    </Grid>
  </Grid>
</Window>
```
I debugged `FrameworkElement.TryFindResource`, which is what your `Model` dependency property is calling to try and find the view manager. Unfortunately, since the library is optimized it's not possible for me to inspect variable states and such, but I am nearly certain `TryFindResource` is returning `null` for some reason. I don't know enough about WPF's internals to know if there is some disconnect due to my use of data templates and `ContentControl`.

**Comment by nathan-alden-hp** [comment]
I quick-watched the `d` variable in `View.ModelProperty`. Sure enough, d.Parent is the `ItemsPanelTemplate` `StackPanel` and `d.Parent.Parent` is `null`. It seems that the way Stylet tries to find the view manager may not be compatible with bindings within data templates.

I'd guess WPF is constructing the data template instance and then only after finishing construction does it add it to the view hierarchy.

**Comment by nathan-alden-hp** [comment]
Since it's likely relevant, here's the definition of the `Modules` property being bound to the `ItemsControl`:
```cs
public BindableCollection<NavigationItemViewModel> Modules { get; } = new BindableCollection<NavigationItemViewModel>();
```
and `NavigationItemViewModel`:
```cs
public class NavigationItemViewModel : Screen
{
    private readonly IMessageQueue _messageQueue = null!;
    private bool _isSelected;

    public NavigationItemViewModel(IMessageQueue messageQueue, IUiModule uiModule, bool isLast)
    {
        _messageQueue = messageQueue;

        SdkIdentifier = uiModule.SdkIdentifier;
        Symbol = uiModule.NavigationSymbol;
        Text = uiModule.NavigationText;
        Visibility = isLast ? Visibility.Collapsed : Visibility.Visible;
    }

    public SdkIdentifier SdkIdentifier { get; }
    public char Symbol { get; }
    public string Text { get; }
    public Visibility Visibility { get; }

    public bool IsSelected
    {
        get => _isSelected;
        set
        {
            _isSelected = value;
            NotifyOfPropertyChange();
        }
    }

    public void Select()
    {
        _messageQueue.Publish(new UiModuleSelectedMessage(SdkIdentifier));
    }
}
```

**Comment by canton7** [comment]
Nah, the VM's aren't relevant here: it's the views. I've used s:View.Model in DataTemplates all the time, so that's not the issue.

**Comment by canton7** [comment]
Ah, I missed your earlier post about the view. Still, I don't see anything that I haven't done before.

I could always fallback to checking Application.Current if TryFindResource fails. Can you confirm that the IViewManager is in Application.Current.Resources?

**Comment by nathan-alden-hp** [comment]
I did some checking and the way I'm using `ItemsControl` appears to be pretty standard. I had been thinking perhaps I wasn't doing something idiomatically. That means there's not much I can change on my side. Any improvement you could make for this scenario would be greatly appreciated.

Here's a screenshot of me inspecting the `viewManager` variable that shows that view manager is available where you said it should be.
![image](https://user-images.githubusercontent.com/40613229/75995420-663e8700-5eca-11ea-8402-d7636e3ff1de.png)

**Comment by canton7** [comment]
Can you confirm that the IViewManager is in Application.Current.Resources, if you have a breakpoint in View when it's looking for the IViewManager?

**Comment by nathan-alden-hp** [comment]
![image](https://user-images.githubusercontent.com/40613229/76031188-36f93b80-5f05-11ea-9a34-c0733833902c.png)

**Comment by canton7** [comment]
Weird. I'll try and reproduce your situation later today. I don't suppose you made any changes to that XAML around the time that this issue started occurring?

There's a NuGet feed at https://ci.appveyor.com/nuget/stylet, mind trying version 1.3.1.196?

**Comment by nathan-alden-hp** [comment]
The only thing I did was to have my `NavigationItemViewModel` class inherit from `Screen` rather than `object`. The view was unchanged.

**Comment by nathan-alden-hp** [comment]
Well now, I feel dumb...

I figured out the exact cause. First of all, Stylet's code is fine. The issue is that I am loading these plugin assemblies at runtime, but _I'm loading the assembly **and its dependencies** into a different `AssemblyLoadContext`_. That means this line's `TryFindResource` in your `View` class:
```cs
var viewManager = ((FrameworkElement)d).TryFindResource(ViewManagerResourceKey) as IViewManager;
```
is working, but the `as IViewManager` is returning `null` because your code is executing in a separate `AssemblyLoadContext` than the main application. This means there are effectively *two* `IViewManager` interfaces. The safe cast fails because `SkinAwareViewManager` implements the `IViewManager` *in the main application's* `AssemblyLoadContext`.

I fixed the problem by defining the `PackageReference` this way in my plugin assembly's project file:
```xml
<PackageReference Include="Stylet" Version="1.3.1">
  <Private>false</Private>
  <ExcludeAssets>runtime</ExcludeAssets>
</PackageReference>
```
I use this technique for project references to avoid the same exact issue, but I neglected to use it for this particular reference.

Anyway, sorry for the goose chase. Needless to say I've learned a good lesson about `AssemblyLoadContext` and `PackageReference`.
